fix: improve `funcId` recognition and squash misdetection bugs

novusnota opened the pull request:
That took a looooooooot of RegEx iterations.

Closes #635

<!--
IMPORTANT:
If your PR doesn't close a particular issue, please, create the issue first and describe the whole context: what you're adding/changing and why you're doing so. And only then open the Pull Request, which would close that issue!

In case you are adding a new language feature, a standard library function or introducing other user-facing changes, you need to document them via a new PR to tact-docs.
-->

- [x] I have updated CHANGELOG.md
~~- [ ] I have documented my contribution in Tact Docs: https://github.com/tact-lang/tact-docs/pull/PR-NUMBER~~
- [x] I have added tests to demonstrate the contribution is correctly implemented: this usually includes both positive and negative tests, showing the happy path(s) and featuring intentionally broken cases
- [x] I have run all the tests locally and no test failure was reported
- [x] I have run the linter, formatter and spellchecker
- [x] I did not do unrelated and/or undiscussed refactorings

novusnota committed:
fix: improve `funcId` recognition and squash misdetection bugs

That took a looooooooot of RegEx iterations, including going into
`grammar.ts` and trying to adjust things in the semantic analysis.
Fortunately, that wasn't needed.

novusnota committed:
chore: CHANGELOG

novusnota committed:
test: chaos ensues

anton-trunov commented on CHANGELOG.md:
```diff
@@ -15,6 +15,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Fixed
 
+- FunC function identifiers with characters from hexadecimal set: PR [#635](https://github.com/tact-lang/tact/pull/635)
```
```suggestion
- FunC function identifiers with characters from hexadecimal set: PR [#636](https://github.com/tact-lang/tact/pull/636)
```

anton-trunov commented on src/grammar/test/items-native-fun-funcid.tact:
```diff
@@ -51,3 +51,38 @@ native idTest17();
+native idTest23();
+
+// NOTE(novusnota):
+// There is a only small set of cases related to the above that our parser won't parse, but C++ FunC's will:
```
looks like this note should go to grammar.ohm

anton-trunov commented:
To be on the safe side, let's take `stdlib.fc` and `mathlib.fc` as libraries that are commonly used and see if we can add more function identifiers from those libs into our positive tests to make sure people will be able to FFI with those.

novusnota committed:
Update CHANGELOG.md

Co-authored-by: Anton Trunov <anton@ton.org>

novusnota committed:
feat/test: extended the possibilities and added more tests

Came up with a simple solution for 0x0_ stuff, so our parser can deal
with them too!

anton-trunov commented on src/grammar/grammar.ohm:
```diff
@@ -263,12 +263,22 @@ Tact {
+
+    /*
+      NOTE(novusnota):
+
```
```suggestion
```

anton-trunov replied:
Let's remove this part, there is commit history to query authorship and additional context

novusnota committed:
Update src/grammar/grammar.ohm

Co-authored-by: Anton Trunov <anton@ton.org>

novusnota committed:
test: corrupted and emojified identifiers

anton-trunov reviewed (changes requested):
This PR accepts, for instance, `+` as a valid FunC identifier, but it's actually not.

Something like this should work:
```
    funcInvalidId = "_" ")" --notUnderscore
                  | ("+" | "-" | "*" | "/" | "~/" | "^/" | "%" | "~%" | "^%" | "/%") ")" --notArithOperator
      // more operators here
                  | ("-"? digit+) ")" --notDecimalNumber
                  | ("-"? "0x" hexDigit+) ")" --notHexadecimalNumber
    funcPlainId = ~funcInvalidId (~(whiteSpace | "(" | ")" | "," | "." | ";" | "~") any)+
```

novusnota commented:
Nice observation! I'd try putting lookaheads `&` in front of those ")" so they don't get parsed, but let me try your suggestion first, it should work. UPD: Yeah, it does, clever idea!

The keywords are not allowed and should be excluded too. Basically, contents of https://github.com/ton-blockchain/ton/blob/master/crypto/func/keywords.cpp has to be put in the exceptions list. I'm on it.

UPD2: Order of operators in those inner () alternations really does matter, the ones with bigger prefix shall stand in front of others.

novusnota committed:
fix/test: prohibit keywords, operators and compiler directives

novusnota committed:
chore: update snapshots

novusnota commented:
<details>
<summary>There was a suggestion, which didn't pass negative tests, so I hid it.</summary>

The thing we've engineered works, but _possibly_ I've found a cleaner solution, which doesn't require the ")" and can be used in the FunC parser too (the current approach in this PR fails there in absence of ")" and stuff). But this one works:

![image](https://github.com/user-attachments/assets/6fe47d83-52e3-442b-a70b-504a7deb14a2)

It passes similar set of tests we have for function identifiers, but adapted to FunC code:

![image](https://github.com/user-attachments/assets/61e7a1f9-487a-4996-8764-c994b05dbee5)

UPD: Negative tests failed terribly, reverting :)

</details>

novusnota committed:
test: derivatives of reserved words should parse

novusnota committed:
fix: typo in the operator

novusnota committed:
fix: prohibit square brackets (used in tuples)

anton-trunov commented:
@novusnota So, what's the current status of this PR?

novusnota commented:
@anton-trunov It's a working and robust solution. The alternative would be to do invalidation of identifiers during syntactic analysis, which enhances the error messages further, but that seems to be a bit out of scope of this PR. Wdyt?
